src/org/exist/storage/XQueryPool.java

```diff
@@ -58,6 +56,8 @@
 
     private final static Logger LOG = LogManager.getLogger(XQueryPool.class);
 
+    private ConcurrentMap<Source, Deque<CompiledXQuery>> pool = new ConcurrentHashMap<>();
+
     private long lastTimeOutCheck;
     private long lastTimeOfCleanup;
 
```

Comment:
@shabanovd `returnObject` looks like this, there is an invariant on the `lastTimeOfCleanup` `AtomicLong`:

```java
 private void returnObject(final Source source, final CompiledXQuery xquery) {
        final long ts = source.getCacheTimestamp();
        if (ts == 0 || ts > lastTimeOfCleanup.get()) {
            if (pool.size() >= maxPoolSize) {
                timeoutCheck();
            }
...
```

 Which then calls `timeoutCheck()`, which has an invariant on the`lastTimeOutCheck` `AtomicLong`:

```java
private void timeoutCheck() {
        if (timeoutCheckInterval < 0L) {
            return;
        }

        final long currentTime = System.currentTimeMillis();
        if (currentTime - lastTimeOutCheck.get() < timeoutCheckInterval) {
            return;
```

So you now have two variables, which participate in the same invariant, for which access to each is independently synchronized, where as previously they were both within the same synchronized access block.

As I mentioned, this may be what you intended, but it changes the concurrent consistency guarantees of the class, so I just wanted to make sure, that that was what you had indeed intended?